### PR TITLE
feat(core): role-scope api

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Add the following API:

- `GET /roles/:id/scopes`
- `POST /roles (with scopeIds)`
- `POST /roles/:id/scopes`
- `DELETE /roles/:id/scopes/:scopeId`

## Testing

UTs

### CODE REVIEW

File: packages/phrases/src/locales/de/errors.ts [hunk 1]
```diff
@@ -177,6 +177,7 @@ const errors = {
   },
   role: {
     name_in_use: 'This role name is already in use', // UNTRANSLATED
+    scope_exists: 'The scope id {{scopeId}} is already been added to this role', // UNTRANSLATED
```
Comment (charIeszhao) [inline packages/phrases/src/locales/de/errors.ts:180]:
```suggestion
    scope_exists: 'The scope id {{scopeId}} has already been added to this role', // UNTRANSLATED
```

File: packages/core/src/routes/role.ts [hunk 2]
```diff
@@ -27,21 +34,28 @@ export default function roleRoutes<T extends AuthedRouter>(router: T) {
...
+        ...roleBody,
         id: roleId(),
       });
 
+      if (scopeIds) {
+        await Promise.all(scopeIds.map(async (scopeId) => findScopeById(scopeId)));
+        await insertRolesScopes(scopeIds.map((scopeId) => ({ roleId: role.id, scopeId })));
+      }
+
+      ctx.body = role;
```
Comment (darcyYe) [inline packages/core/src/routes/role.ts:57]: we do not have `scopeIds` in `ctx.body`, is this expected?

Comment (darcyYe) [reply packages/core/src/routes/role.ts:57]: Seems yes, aligned with other role APIs.

File: packages/core/src/routes/role.ts [hunk 3]
```diff
@@ -98,4 +112,72 @@ export default function roleRoutes<T extends AuthedRouter>(router: T) {
...
+        params: { id },
+        body: { scopeIds },
+      } = ctx.guard;
+
+      await findRoleById(id);
+      const rolesScopes = await findRolesScopesByRoleId(id);
+
+      for (const scopeId of scopeIds) {
+        assertThat(
+          !rolesScopes.some(({ scopeId: _scopeId }) => _scopeId === scopeId),
```
Comment (darcyYe) [inline packages/core/src/routes/role.ts:151]: ditto

File: packages/core/src/routes/role.ts [hunk 4]
```diff
@@ -27,21 +34,28 @@ export default function roleRoutes<T extends AuthedRouter>(router: T) {
   router.post(
     '/roles',
     koaGuard({
-      body: Roles.createGuard.omit({ id: true }),
+      body: Roles.createGuard
+        .omit({ id: true })
+        .extend({ scopeIds: z.string().min(1).array().optional() }),
```
[the remark this answers is not in the document]
Comment (wangsijie) [reply packages/core/src/routes/role.ts:39]: Our API serves Admin Console first, currently no such design.

File: packages/core/src/routes/role.ts (hunk 3 above)
Comment (wangsijie) [reply packages/core/src/routes/role.ts:151]: Can you please explain more details?

File: packages/core/src/queries/roles-scopes.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  `);
+
+export const deleteRolesScope = async (roleId: string, scopeId: string) => {
+  const { rowCount } = await envSet.pool.query(sql`
+    delete from ${table}
+    where ${fields.scopeId} = ${scopeId} and ${fields.roleId} = ${roleId}
+  `);
+
+  if (rowCount < 1) {
+    throw new DeletionError(RolesScopes.table);
```
[the remark this answers is not in the document]
Comment (wangsijie) [reply packages/core/src/queries/roles-scopes.ts:34]: `DeletionError` only supports one id, but there are 2 ids here, created LOG-5085 to optimize.

File: packages/phrases/src/locales/de/errors.ts (hunk 1 above)
Comment (wangsijie) [reply packages/phrases/src/locales/de/errors.ts]: updated

**Review by simeng-li — APPROVED** [review]
overall LGTM

File: packages/core/src/routes/role.ts [hunk 6]
```diff
@@ -27,21 +34,28 @@ export default function roleRoutes<T extends AuthedRouter>(router: T) {
...
     async (ctx, next) => {
-      const {
-        body,
-        body: { name },
-      } = ctx.guard;
+      const { body } = ctx.guard;
+      const { scopeIds, ...roleBody } = body;
 
-      assertThat(!(await findRoleByRoleName(name)), 'role.name_in_use');
+      assertThat(!(await findRoleByRoleName(roleBody.name)), 'role.name_in_use');
```
[the remark this answers is not in the document]
Comment (wangsijie) [reply packages/core/src/routes/role.ts]: updated

**Review by darcyYe — APPROVED** [review]

**Pull request merged**